source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc

```diff
@@ -451,46 +451,50 @@ Http::FilterTrailersStatus JsonTranscoderFilter::encodeTrailers(Http::HeaderMap&
 
   response_in_.finish();
 
-  Buffer::OwnedImpl data;
-  readToBuffer(*transcoder_->ResponseOutput(), data);
-
-  if (data.length()) {
-    encoder_callbacks_->addEncodedData(data, true);
-    has_body_ = true;
-  }
-
-  if (method_->server_streaming()) {
-    // For streaming case, the headers are already sent, so just continue here.
-    return Http::FilterTrailersStatus::Continue;
-  }
-
   // If there was no previous headers frame, this |trailers| map is our |response_headers_|,
   // so there is no need to copy headers from one to the other.
   bool is_trailers_only_response = response_headers_ == &trailers;
 
   const absl::optional<Grpc::Status::GrpcStatus> grpc_status =
       Grpc::Common::getGrpcStatus(trailers);
   bool status_converted_to_json = grpc_status && maybeConvertGrpcStatus(*grpc_status, trailers);
+  if (status_converted_to_json) {
```

Comment:
The code is very hard to follow with so many if-else.  Can we split them into two functions, one handling all logic of convert_to_json=true case, and another handling false case.  Not need to worry about sharing code between them.